pkg/fqdn/cache.go

```diff
@@ -699,6 +746,105 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
 	return nil
 }
 
+// UpdateVirtualIPs updates the mapping from frontends to backends as stored in the cache.
+// Returns the set of names that reference these frontends.
+func (c *DNSCache) UpdateVirtualIPs(frontends []netip.Addr, backends []netip.Addr, mergeExisting bool) []string {
+	c.Lock()
+	defer c.Unlock()
+	v4be, v6be := ippkg.SplitAddrsByFamily(backends)
+
+	frontendsWithChanges := sets.Set[netip.Addr]{}
+	for _, frontend := range frontends {
+		changed := false
+		if frontend.Is4() {
+			changed = c.setVirtualIP(frontend, v4be, mergeExisting)
+		} else {
+			changed = c.setVirtualIP(frontend, v6be, mergeExisting)
+		}
+		if changed {
+			frontendsWithChanges.Insert(frontend)
+		}
+	}
+
+	// If nothing changed; we're done
+	if len(frontendsWithChanges) == 0 {
+		return nil
+	}
+
+	updatedNames := []string{}
+	for name, entry := range c.forward {
+		for ip := range entry {
+			if frontendsWithChanges.Has(ip) {
```

Comment:
This loop looks kind of expensive, and we need to call it on every single service update. We may wish to consider another tracking map instead.